mitre/caldera
# Revamp unit testing infrastructure

* Replaces barebones Travis CI configuration with Tox for more in-depth testing featureset and consistency across multiple dev environments
Previously, you would manually install flake8 and use unittest manually to run tests. Tox builds entire virtualenvs with the required dependencies to run tests to mitigate the potential differences between developer environments.

* Switch out python `unittest` for `pytest`
* Add development requirements file
* Add code coverage with `coverage.py` through https://codecov.io/gh/mitre/caldera
* Add [`pre-commit`](https://pre-commit.com/) configuration file that can be used by developers to catch potential errors before committed. It currently only runs flake8 and flake8-bugbear, but it will be incorporating the following `pre-commit` hooks in the future:
```yaml
  - repo: https://github.com/pre-commit/pre-commit-hooks
    rev: v2.2.3
    hooks:
      - id: check-byte-order-marker
      - id: trailing-whitespace
        args: [--markdown-linebreak-ext=md]
      - id: end-of-file-fixer
```
The hooks do the following:
### `check-byte-order-marker`

Forbid files which have a UTF-8 byte-order marker

### `trailing-whitespace`
Trims trailing whitespace.

### `end-of-file-fixer`
Makes sure files end in a newline and only a newline.
